smolyay/surrogate.py

```diff
@@ -1,112 +1,67 @@
+    :attr:`data` stores the output of the true function at the sampled points.
+    :attr:`num_dimensions` is the number of dimensionns/independent variables.
+    :meth:`train`, generates a trained surrogate model given a function and
+    the points to sample at.
+    :meth:`fit` computes the coefficients based on a set of data
+    (function values at transformed grid points).
     Once the surrogate is constructed, one can evaluate the surrogate
-    and its derivative through :meth:`__call__(x)` and :meth:`derivative(x)`
-    at a given input. :meth:`gradient` takes the gradient of the
-    surrogate.
-    :meth:`_reset_surrogate` resets the surrogate's coefficients,
-    sampled points, original ``grids``(holds the grid points
-    and their corresponding basis functions depending on the
-    dimensionality and the generator), and ``data`` (function at
-    sampling points).
+    through :meth:`predict(x)`.
 
     Parameters
     ----------
-    domain: list
-        Domain of the function to be approximated.
-    grid_generator: IndexGridGenerator
-        Set of grids and their corresponding basis function.
-
-    Raises
-    ------
-    TypeError
-        Grids must be :class:`IndexGridGenerator`.
-        Domain must be dim x 2.
-
-    Example
-    -------
-    Consider a function :math:F(y) with one independent variable.
-    Basis function B_n(x) is chosen for the approximation and
-    three grid points and their corresponding basis functions
-    are generated in (-1, 1) domain:
-    (:math:x_0, :math:x_1, :math:x_2)
-    [:math:B_0(x), :math:B_1(x) :math:B_2(x)]
-    Based on the domain of the approximation, grid points of the basis
-    functions can be moved to the new domain,
-    resulting in transformed grid points:
-    (:math:y_0, :math:y_1, :math:y_2)
-    Then:
-    ..math::
-        \[
-        \begin{bmatrix}
-        B_0(x_0)& B_1(x_0) & B_2(x_0) \\
-        B_0(x_1)& B_1(x_1) & B_2(x_1) \\
-        B_0(x_2)& B_1(x_2) & B_2(x_2)
-        \end{bmatrix}
-        \begin{bmatrix}
-        C_0\\C_1\\C_2
-        \end{bmatrix}
-        =\begin{bmatrix}
-        F(y_0)\\F(y_1)\\F(y_2)
-        \end{bmatrix}
-        \]
-
-    Coefficients (x) can be obtained by solving linear equations,
-    since we have a full rank system of matrixes. The user can
-    select between 'lower upper decomposition' ('lu'), and
-    'inverse' ('inv') method for calculating the surrogate's (S)
-    coefficients. 'lstsq' solves the above equations through
-    least linear method.
-    S = C_0B_0(x) + C_1B_1(x) + C_2B_2(x)
+    domain: numpy.ndarray
+        the domain of the surrgoate
     """
 
-    def __init__(self, domain, grid_generator):
-        self.domain = domain
-        self.grid_generator = grid_generator
-        self._coefficients = None
-        self._grid = None
-        self._points = None
+    def __init__(self, domain):
+        self._domain = None
         self._data = None
+        self._points = None
```

Comment:
Why do we need the points in this class? Those would only used by the fit method for now, so I wouldn't save them.

Reply:
Possibly for retraining the data after changing certain settings for the regression. Also for knowing what points were used for the regression in the first place which is important for the optimization workflow.

Reply:
The user knows what points they used for training because they had the point set, so remove this.

Reply:
The GaussianProcessRegressor also keeps a copy (or a reference depending on whether copy_X_train is true) of the training variables for safekeeping. They're called X_train_ and y_train_.
Also, if I were to train a model and save it to a file, then open that file in another script, then I won't have the point set. That would be a huge problem for the optimization workflow and I'd need to generate the point set all over again.

I don't think removing them is a good idea, but I think renaming them to X_train_ and y_train_ to match sklearn is a good idea.

Reply:
Your logic is circuitous: if you don't keep the points in this class, you don't need them to reconstruct the model from a file (only the coefficients). We also don't have a mechanism for doing this currently anyway, and we should discuss how to do that separately if we want to enable it. As we discussed, we are also **not** trying to reproduce a sklearn interface, just to have an object that is compatible with their philosophy.

Please remove this.

Reply:
> Your logic is circuitous: if you don't keep the points in this class, you don't need them to reconstruct the model from a file (only the coefficients).

What? The points aren't used to reconstruct the model. I'm referring to the fact that in the workflow, after I train a model I save it to a pickle file that I can later open in a different script. The points saved as a model property are not used to reconstruct the model. The model is already constructed. They're there so I can know what points were used to train the model without having to look at the job statepoint and reconstruct the entire point set all over again. The optimization workflow needs them because those points are candidate initial guesses, and it's really convenient for the model to just have those points available.

A lot of my analysis of these surrogates boils down to comparing different performance stats vs the number of points the model was trained, and that gets a lot more cumbersome if I can't tell just from the model's properties how many points it was trained on. With the number of terms decoupled from the number of training points, it's going to get very frustrating if I can't easily get access to the training points, or just the number of training points. if `points` and `data` are going to be removed, at least let it save how many sample points there were.

Reply:
You can save the training points for a model yourself if you need them for your workflow. It would look like:
```
X = UniformRandomPointSet(...)
y = f(X)
np.savetxt("training_data.dat", np.column_stack(X.points, y))

model.fit(X, y)
```

Reply:
It's removed, but this kills any potential ability for the surrogate to self-assess how good it is at fitting to its own samples outside of `fit` and `fit_gradient`.

Reply:
That is easily done:
```py
model.fit(X, y)
np.linalg.norm(model.predict(X) - y)
```
You would do the same thing if you had test data that wasn't in the training set.

This could be done internally during fitting if it was important.

Reply:
I will say that the memory reduction by removing them is pretty huge.